<img src='https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcQ-VfNtOyJbsaxu43Kztf_cv1mgBG6ZIQZEVw&usqp=CAU'>

# Procesamiento de Lenguaje Natural

## Taller # 5

## Yofre Alexander Sanabria Arias

In [351]:
import re
import numpy as np
import pandas as pd

from nltk.corpus import stopwords
stopwords_sp = stopwords.words('spanish')

from nltk.stem.snowball import SnowballStemmer
spanishStemmer=SnowballStemmer("spanish")

from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_distances

# Punto 1: Pre-Procesamiento

In [352]:
data_bob=pd.DataFrame(pd.read_csv('bob_esponja.csv'))
data_bob

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  
0  Bob Esponja es una esponja de mar con forma re...  
1  Es una estrella de mar de color rosa, el mejor...  
2  Calamardo Tentáculos es un personaje principal...  
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...  
4  Don Eugenio H. Cangrejo (normalmente llamado S...  
5  Sheldon J. Plankton, o simplemente Plankton, e...  
6  Karen Plankton es uno de los dos principales a...  
7  Perla "Perlita" Cangrejo es un personaje princ...  
8  La Señora Puff es un personaje principal de Bo...  
9  Gary el Caracol es un personaje principal de B...

# Punto 2: TF-IDF

In [353]:
data_bob['pre-procesado']=""

In [354]:
cadena=""
for i in range(0, len(data_bob['Descripción'])):
    dep=data_bob['Descripción'][i] # Guardar el contenido de la celda en la variable dep
    dep=dep.lower() # Comverti a minuscula
    #dep=re.sub(r"[^a-zA-Z\s\d]", " ", dep)  
    dep=re.sub(r"[^a-zA-Záéíóúñ\s]", " ", dep) # Dejar solo letras
    dep=dep.split() # crear array con las palabras
    texto=[palabra for palabra in dep if palabra not in stopwords_sp] # Eliminar palabras vacias 
    Cadena=""
    for ele in texto:
        ele=spanishStemmer.stem(ele)
        espacio=" "
        cadena += espacio
        cadena += ele       
    data_bob['pre-procesado'][i]=cadena
    cadena=""

In [355]:
data_bob

Personaje  \
0  Bob Esponja Pantalones Cuadrados   
1                 Patricio Estrella   
2              Calamardo Tentáculos   
3                  Arenita Mejillas   
4               Eugenio H. Cangrejo   
5               Sheldon J. Plankton   
6                    Karen Plankton   
7                  Perlita Cangrejo   
8                         Sra. Puff   
9                   Gary el Caracol   

                                         Descripción  \
0  Bob Esponja es una esponja de mar con forma re...   
1  Es una estrella de mar de color rosa, el mejor...   
2  Calamardo Tentáculos es un personaje principal...   
3  Arenita Mejillas (Sandy Cheeks en inglés) en u...   
4  Don Eugenio H. Cangrejo (normalmente llamado S...   
5  Sheldon J. Plankton, o simplemente Plankton, e...   
6  Karen Plankton es uno de los dos principales a...   
7  Perla "Perlita" Cangrejo es un personaje princ...   
8  La Señora Puff es un personaje principal de Bo...   
9  Gary el Caracol es un personaje principal de B...   

                                       pre-procesado  
0   bob esponj esponj mar form rectangul color am...  
1   estrell mar color ros mejor amig bob esponj j...  
2   calamard tentacul personaj principal bob espo...  
3   arenit mejill sandy cheeks ingles personaj pr...  
4   don eugeni h cangrej normal llam señor cangre...  
5   sheldon j plankton simplement plankton dos an...  
6   kar plankton dos principal antagon bob esponj...  
7   perl perlit cangrej personaj principal bob es...  
8   señor puff personaj principal bob esponj maes...  
9   gary caracol personaj principal bob esponj qu...

# Punto 3: Distancia del coseno

In [356]:
index=list(data_bob['Personaje'])

tfidf_vect = TfidfVectorizer()
tfidf = tfidf_vect.fit_transform(data_bob['pre-procesado'].values)
tfidf_matrix = pd.DataFrame(tfidf.toarray(), columns=tfidf_vect.get_feature_names())
tfidf_matrix.index = index

tfidf_matrix.T.round(3)

Bob Esponja Pantalones Cuadrados  Patricio Estrella  \
abaj                                 0.057               0.00   
aborrec                              0.000               0.00   
acab                                 0.000               0.00   
accident                             0.049               0.00   
acept                                0.000               0.00   
...                                    ...                ...   
viv                                  0.000               0.05   
vol                                  0.057               0.00   
voz                                  0.195               0.00   
vuelv                                0.172               0.00   
zapat                                0.043               0.00   

          Calamardo Tentáculos  Arenita Mejillas  Eugenio H. Cangrejo  \
abaj                     0.000             0.000                0.000   
aborrec                  0.091             0.000                0.000   
acab                     0.091             0.000                0.000   
accident                 0.000             0.000                0.045   
acept                    0.000             0.000                0.000   
...                        ...               ...                  ...   
viv                      0.060             0.108                0.070   
vol                      0.000             0.000                0.000   
voz                      0.000             0.139                0.000   
vuelv                    0.000             0.000                0.000   
zapat                    0.000             0.000                0.000   

          Sheldon J. Plankton  Karen Plankton  Perlita Cangrejo  Sra. Puff  \
abaj                      0.0           0.000               0.0      0.000   
aborrec                   0.0           0.000               0.0      0.000   
acab                      0.0           0.000               0.0      0.000   
accident                  0.0           0.000               0.0      0.000   
acept                     0.0           0.064               0.0      0.000   
...                       ...             ...               ...        ...   
viv                       0.0           0.000               0.0      0.000   
vol                       0.0           0.000               0.0      0.000   
voz                       0.0           0.000               0.0      0.000   
vuelv                     0.0           0.000               0.0      0.000   
zapat                     0.0           0.000               0.0      0.093   

          Gary el Caracol  
abaj                0.000  
aborrec             0.000  
acab                0.000  
accident            0.000  
acept               0.000  
...                   ...  
viv                 0.000  
vol                 0.000  
voz                 0.000  
vuelv               0.000  
zapat               0.068  

[709 rows x 10 columns]

In [357]:
dist_cos = cosine_distances(tfidf_matrix.values)
dist_cos = pd.DataFrame(dist_cos, columns = tfidf_matrix.index, index = tfidf_matrix.index)
dist_cos

Bob Esponja Pantalones Cuadrados  \
Bob Esponja Pantalones Cuadrados                          0.000000   
Patricio Estrella                                         0.826466   
Calamardo Tentáculos                                      0.918853   
Arenita Mejillas                                          0.838917   
Eugenio H. Cangrejo                                       0.868554   
Sheldon J. Plankton                                       0.872613   
Karen Plankton                                            0.938901   
Perlita Cangrejo                                          0.886072   
Sra. Puff                                                 0.857937   
Gary el Caracol                                           0.924381   

                                  Patricio Estrella  Calamardo Tentáculos  \
Bob Esponja Pantalones Cuadrados           0.826466              0.918853   
Patricio Estrella                          0.000000              0.791198   
Calamardo Tentáculos                       0.791198              0.000000   
Arenita Mejillas                           0.841713              0.920036   
Eugenio H. Cangrejo                        0.831869              0.831488   
Sheldon J. Plankton                        0.966669              0.917784   
Karen Plankton                             0.930557              0.950219   
Perlita Cangrejo                           0.891881              0.967358   
Sra. Puff                                  0.952574              0.968076   
Gary el Caracol                            0.895000              0.852161   

                                  Arenita Mejillas  Eugenio H. Cangrejo  \
Bob Esponja Pantalones Cuadrados          0.838917             0.868554   
Patricio Estrella                         0.841713             0.831869   
Calamardo Tentáculos                      0.920036             0.831488   
Arenita Mejillas                          0.000000             0.908027   
Eugenio H. Cangrejo                       0.908027             0.000000   
Sheldon J. Plankton                       0.967723             0.687797   
Karen Plankton                            0.948562             0.892102   
Perlita Cangrejo                          0.962264             0.796912   
Sra. Puff                                 0.972093             0.895877   
Gary el Caracol                           0.900968             0.920780   

                                  Sheldon J. Plankton  Karen Plankton  \
Bob Esponja Pantalones Cuadrados             0.872613        0.938901   
Patricio Estrella                            0.966669        0.930557   
Calamardo Tentáculos                         0.917784        0.950219   
Arenita Mejillas                             0.967723        0.948562   
Eugenio H. Cangrejo                          0.687797        0.892102   
Sheldon J. Plankton                          0.000000        0.795121   
Karen Plankton                               0.795121        0.000000   
Perlita Cangrejo                             0.922782        0.966336   
Sra. Puff                                    0.986743        0.984939   
Gary el Caracol                              0.969443        0.965513   

                                  Perlita Cangrejo  Sra. Puff  Gary el Caracol  
Bob Esponja Pantalones Cuadrados          0.886072   0.857937         0.924381  
Patricio Estrella                         0.891881   0.952574         0.895000  
Calamardo Tentáculos                      0.967358   0.968076         0.852161  
Arenita Mejillas                          0.962264   0.972093         0.900968  
Eugenio H. Cangrejo                       0.796912   0.895877         0.920780  
Sheldon J. Plankton                       0.922782   0.986743         0.969443  
Karen Plankton                            0.966336   0.984939         0.965513  
Perlita Cangrejo                          0.000000   0.873928         0.969604  
Sra. Puff                                 0.873928   0.0

In [358]:
dist_cos=dist_cos.round(3)

# ¿Cuáles son los personajes menos parecidos?
- Los personajes menos parecidos son la __Sra. Puff y Sheldon J. Plankton__ con un valor de 0.987 	

In [359]:
Menos_parecidos = dist_cos[dist_cos == np.amax(dist_cos.values)] # Obtener el mayor valor en la matriz
Menos_parecidos

Bob Esponja Pantalones Cuadrados  \
Bob Esponja Pantalones Cuadrados                               NaN   
Patricio Estrella                                              NaN   
Calamardo Tentáculos                                           NaN   
Arenita Mejillas                                               NaN   
Eugenio H. Cangrejo                                            NaN   
Sheldon J. Plankton                                            NaN   
Karen Plankton                                                 NaN   
Perlita Cangrejo                                               NaN   
Sra. Puff                                                      NaN   
Gary el Caracol                                                NaN   

                                  Patricio Estrella  Calamardo Tentáculos  \
Bob Esponja Pantalones Cuadrados                NaN                   NaN   
Patricio Estrella                               NaN                   NaN   
Calamardo Tentáculos                            NaN                   NaN   
Arenita Mejillas                                NaN                   NaN   
Eugenio H. Cangrejo                             NaN                   NaN   
Sheldon J. Plankton                             NaN                   NaN   
Karen Plankton                                  NaN                   NaN   
Perlita Cangrejo                                NaN                   NaN   
Sra. Puff                                       NaN                   NaN   
Gary el Caracol                                 NaN                   NaN   

                                  Arenita Mejillas  Eugenio H. Cangrejo  \
Bob Esponja Pantalones Cuadrados               NaN                  NaN   
Patricio Estrella                              NaN                  NaN   
Calamardo Tentáculos                           NaN                  NaN   
Arenita Mejillas                               NaN                  NaN   
Eugenio H. Cangrejo                            NaN                  NaN   
Sheldon J. Plankton                            NaN                  NaN   
Karen Plankton                                 NaN                  NaN   
Perlita Cangrejo                               NaN                  NaN   
Sra. Puff                                      NaN                  NaN   
Gary el Caracol                                NaN                  NaN   

                                  Sheldon J. Plankton  Karen Plankton  \
Bob Esponja Pantalones Cuadrados                  NaN             NaN   
Patricio Estrella                                 NaN             NaN   
Calamardo Tentáculos                              NaN             NaN   
Arenita Mejillas                                  NaN             NaN   
Eugenio H. Cangrejo                               NaN             NaN   
Sheldon J. Plankton                               NaN             NaN   
Karen Plankton                                    NaN             NaN   
Perlita Cangrejo                                  NaN             NaN   
Sra. Puff                                       0.987             NaN   
Gary el Caracol                                   NaN             NaN   

                                  Perlita Cangrejo  Sra. Puff  Gary el Caracol  
Bob Esponja Pantalones Cuadrados               NaN        NaN              NaN  
Patricio Estrella                              NaN        NaN              NaN  
Calamardo Tentáculos                           NaN        NaN              NaN  
Arenita Mejillas                               NaN        NaN              NaN  
Eugenio H. Cangrejo                            NaN        NaN              NaN  
Sheldon J. Plankton                            NaN      0.987              NaN  
Karen Plankton                                 NaN        NaN              NaN  
Perlita Cangrejo                               NaN        NaN              NaN  
Sra. Puff                                      NaN      

# ¿Cuáles son los personajes más parecidos?
- Los personajes más parecidos son __Eugenio H. Cangrejo y Sheldon J. Plankton__ con un valor de 0.688

In [360]:
dist_cos.values[dist_cos.values==0]=2 #Reemplazar lo valores de 0 por un 2

In [361]:
Mas_parecidos = dist_cos[dist_cos == np.amin(dist_cos.values)]# Buscar el valor menor en la matriz
Mas_parecidos

Bob Esponja Pantalones Cuadrados  \
Bob Esponja Pantalones Cuadrados                               NaN   
Patricio Estrella                                              NaN   
Calamardo Tentáculos                                           NaN   
Arenita Mejillas                                               NaN   
Eugenio H. Cangrejo                                            NaN   
Sheldon J. Plankton                                            NaN   
Karen Plankton                                                 NaN   
Perlita Cangrejo                                               NaN   
Sra. Puff                                                      NaN   
Gary el Caracol                                                NaN   

                                  Patricio Estrella  Calamardo Tentáculos  \
Bob Esponja Pantalones Cuadrados                NaN                   NaN   
Patricio Estrella                               NaN                   NaN   
Calamardo Tentáculos                            NaN                   NaN   
Arenita Mejillas                                NaN                   NaN   
Eugenio H. Cangrejo                             NaN                   NaN   
Sheldon J. Plankton                             NaN                   NaN   
Karen Plankton                                  NaN                   NaN   
Perlita Cangrejo                                NaN                   NaN   
Sra. Puff                                       NaN                   NaN   
Gary el Caracol                                 NaN                   NaN   

                                  Arenita Mejillas  Eugenio H. Cangrejo  \
Bob Esponja Pantalones Cuadrados               NaN                  NaN   
Patricio Estrella                              NaN                  NaN   
Calamardo Tentáculos                           NaN                  NaN   
Arenita Mejillas                               NaN                  NaN   
Eugenio H. Cangrejo                            NaN                  NaN   
Sheldon J. Plankton                            NaN                0.688   
Karen Plankton                                 NaN                  NaN   
Perlita Cangrejo                               NaN                  NaN   
Sra. Puff                                      NaN                  NaN   
Gary el Caracol                                NaN                  NaN   

                                  Sheldon J. Plankton  Karen Plankton  \
Bob Esponja Pantalones Cuadrados                  NaN             NaN   
Patricio Estrella                                 NaN             NaN   
Calamardo Tentáculos                              NaN             NaN   
Arenita Mejillas                                  NaN             NaN   
Eugenio H. Cangrejo                             0.688             NaN   
Sheldon J. Plankton                               NaN             NaN   
Karen Plankton                                    NaN             NaN   
Perlita Cangrejo                                  NaN             NaN   
Sra. Puff                                         NaN             NaN   
Gary el Caracol                                   NaN             NaN   

                                  Perlita Cangrejo  Sra. Puff  Gary el Caracol  
Bob Esponja Pantalones Cuadrados               NaN        NaN              NaN  
Patricio Estrella                              NaN        NaN              NaN  
Calamardo Tentáculos                           NaN        NaN              NaN  
Arenita Mejillas                               NaN        NaN              NaN  
Eugenio H. Cangrejo                            NaN        NaN              NaN  
Sheldon J. Plankton                            NaN        NaN              NaN  
Karen Plankton                                 NaN        NaN              NaN  
Perlita Cangrejo                               NaN        NaN              NaN  
Sra. Puff                                      NaN      